### DISCUSSION TITLE
Track changes and update remote Playlist

### DISCUSSION THREAD

**Discussion by iSerganov** [savonet/liquidsoap#4016] [Ideas]
I am using liquidsoap 2.2.5 with remote playlist source

I need to reload the playlist from remote source once it is updated. 
And unfortunately I cannot achieve the above behavior. 

My config is as as follows:

```
#!/usr/bin/liquidsoap
# Log dir
log.file.path.set("/tmp/basic-radio.log")

# Music
myplaylist = playlist("https://some/remote/path/playlist.pls",reload_mode="watch",mode="normal")
myplaylist.on_track(print)
# If something goes wrong, we'll play this
noiseS = noise()

radio = myplaylist
# And finally the security
radio = fallback(track_sensitive = false, [radio, noiseS])

# Stream it out
output.icecast(%mp3,
  host = "host", port = 9550,
  password = "pass", mount = "/radio.mp3",
  radio)

```

I expect that every time I update the remote playlist under the above URL from config (e.g., https://some/remote/path/playlist.pls),
it will will be automatically reloaded due to `reload_mode="watch"` param. However this doesn't happen.

 I also tried the following option: `myplaylist = playlist("https://some/remote/path/playlist.pls",reload=5,mode="normal")`.
 Which indeed reloads playlist every 5 seconds, but that leads to always playing the first 2(?!) songs in the playlist(no matter how long it is).
 
 Can you please advise what am I doing wrong?
 Maybe there is an option to fetch remote playlist every N seconds, compare it with the currently played playlist and reload if any change is detected?
 
 Thank you.

**Comment by gAlleb** [comment]
Hi @iSerganov. Remote playlists don't support `reload_mode="watch"`. That's how it is.

You can reload a playlist via telnet: `myplaylist.reload` or make a harbor function, just paste it after your playlist and then you can reload it via `http://locahost:8008/reload_playlist`

```
harbor.http.register.simple(port=8008, "/reload_playlist", fun(_) -> begin myplaylist.reload(); "myplaylist has been reloaded! Awesome!" end)

**Reply by vitoyucepi** [reply]
Hi @iSerganov,
As explained above, you can call the reload function from the external script by creating an API endpoint.
It's also possible to
- Reload the playlist every N seconds or rounds. A round is a full loop of the playlist.
- Create a custom downloader that uses `http.get` with an `if-modified-since` header. The result could be passed to the `playlist.list` function.

> but that leads to always playing the first 2(?!) songs

You may want to use `playlist(mode="random")` to randomize the playlist in combination with the `playlog` function to drop repetitions.

**Comment by iSerganov** [comment]
Thank you for the fast response and explanation.
I have a few questions/requests then:

- Probably it worth mentioning in the documentation that `reload_mode="watch"` doesn't work with remote playlists.
- How to convert this discussion to a feature request? ;)
- Thank you for the option with `harbor` - it's a good workaround!

There is another workaround (just in case if someone wants automatically updated remote playlist).
Run a small shell script that periodically CURLs the remote playlist and saves it to a file if **newer version is available**:

```shell
while [ true ]
do
    if test -e "/etc/liquidsoap/playlist.pls"
    then zflag="-z /etc/liquidsoap/playlist.pls"
    else zflag=
    fi
    curl -o "/etc/liquidsoap/playlist.pls" $zflag "https://your.remote.domain/playlist.pls"
    sleep <your_interval_in_seconds>
done
```
And liquidsoap playlist definition should be like this: `myplaylist = playlist("/etc/liquidsoap/playlist.pls",reload_mode="watch",mode="normal")`

**Reply by vitoyucepi** [reply]
> How to convert this discussion to a feature request? ;)

I suggest creating a new issue using the feature request template and referencing this discussion.
